### PR TITLE
Initial code import.

### PR DESCRIPTION
**Pull request by carlosalberto**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+from setuptools import setup
+
+version = open('VERSION').read()
+setup(
+    name='redis_opentracing',
+    version=version,
+    url='https://github.com/carlosalberto/python-redis/',
```
Comment (joeblubaugh) [inline setup.py:7]:
```
url
download_url
description
```

all likely need to change when we do a PR against opentracing-contrib.

File: redis_opentracing/__init__.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    # Patch the execute() method.
+    execute_method = pipe.execute
+
+    @wraps(execute_method)
+    def tracing_execute(raise_on_error=True):
+        if not pipe.command_stack:
+            # Nothing to process/handle.
+            return execute_method(raise_on_error=raise_on_error)
+
+        span = g_tracer.start_span(_get_operation_name('MULTI'))
```
Comment (joeblubaugh) [inline redis_opentracing/__init__.py:112]: Is there a reasonable method for attaching a function that will give us the parent_span (if any?)

These calls to redis likely have upstream spans associated with a request that hits the server. We ought to figure out a way to make sure they get joined together with the current_span, or some other span context.

File: examples/explicit.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import datetime
+import redis
+
+import lightstep
```
Comment (joeblubaugh) [inline examples/explicit.py]: Do we want to directly use LightStep in this tracer example? Can we avoid requiring any OT implementation and just show some kind of "set up your tracer" placeholder?

File: redis_opentracing/__init__.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
+from functools import wraps
+
+import redis
+
+g_tracer = None
+g_trace_prefix = None
+g_trace_all_classes = True
+
+def init_tracing(tracer, trace_all_classes=True, prefix='Redis'):
```
Comment (joeblubaugh) [inline redis_opentracing/__init__.py:9]: Can you add doc strings for all the public defs in this file that adhere to [PEP 287](https://www.python.org/dev/peps/pep-0287/)

In particular, clarifying the role of `trace_all_classes` and `prefix` for users would be great.

**Review by joeblubaugh — COMMENTED** [review]
Hi Carlos. I'm Joe, an engineer at LightStep.

At a high level, I think my main question about this code is figuring out a way that we can link these spans with either a parent span from the same process or a remote span whose context has been propagated to this process.

This is a little "up in the air" right now, but both the Django and Flask implementations on opentracing-contrib provide methods to get the span for a given request. Is there a common question interface we can define that will let us access those spans when starting new ones?

File: setup.py (hunk 1 above)
Comment (carlosalberto) [reply setup.py:7]: Oh definitely.

File: redis_opentracing/__init__.py (hunk 4 above)
Comment (carlosalberto) [reply redis_opentracing/__init__.py:9]: Sure, the plan was to add it today at some point (with a good, clear README as well) :)

File: redis_opentracing/__init__.py (hunk 2 above)
Comment (carlosalberto) [reply redis_opentracing/__init__.py:112]: So, the thing about the parent span, I TOTALLY FORGOT to mention as part as the PR: there's a current effort to provide an ActiveSpanSource, which is similar with what happened with Java recently. This is specially important when incorporating tracing with async frameworks (Tornado, gevent, the new asyncio). Once we have something in place, I will add the required hooks this and there.

Observe the other libraries I worked on (elasticsearch, pyramid, sqlalchemy) provide their own parent span calls (mostly on a thread-local basis... which won't be enough - so they will have to be updated eventually, too). Hope that helps :)

File: examples/explicit.py (hunk 3 above)
Comment (carlosalberto) [reply examples/explicit.py]: So far we have been using the `lightstep` tracer as part of the examples everywhere. One thing we could do is simply use the `opentracing.tracer` instance from here, and put a comment on the examples to set that first. 

Else, at least a comment there, saying `Or replace it here the tracer of your choice`.

File: README.rst [hunk 5]
```diff
@@ -0,0 +1,65 @@
+#################
+Redis Opentracing
+#################
+
+This package enables distributed tracing for the Python redis library.
+
+Instalation
```
Comment (joeblubaugh) [inline README.rst]: Instalation -> Installation

File: README.rst [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+===========
+
+Run the following command:
+
+    $ pip install redis_opentracing
+
+Getting started
+===============
+
+Please see the examples directory. Overall, usage requires that a tracer gets set and have the Redis client and pipelines functionality get automatically traced:
```
Comment (joeblubaugh) [inline README.rst]: Maybe rephrase this:

```
Tracing a Redis client requires setting up an OpenTracing-compatible tracer, and calling `init_tracing` to set up the tracing wrappers. See the examples directory for several different approaches.
```

File: README.rst [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    import redis
+    import redis_opentracing
+
+    redis_opentracing.init_tracing(tracer)
+
+    client = redis.StrictRedis()
+    client.set('last_access', datetime.datetime.now())
+
+It's possible to have traced only specific `StrictRedis` client objects:
```
Comment (joeblubaugh) [inline README.rst]: Rephrase: `It's possible to trace only specific Redis clients:`

File: README.rst [hunk 8]
```diff
@@ -0,0 +1,65 @@
+#################
```
Comment (joeblubaugh) [inline README.rst:1]: Great README, thanks!

File: README.rst [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    pipe = client.pipeline()
+    redis_opentracing.trace_pipeline(pipe)
+
+    # This pipeline will be executed as a single MULTI command.
+    pipe.lpush('fruits', 'lemon', 'watermelon')
+    pipe.rpush('fruits', 'pineapple', 'apple')
+    pipe.execute()
+
+
+A thing to notice about pipelines is that when executed as a transaction, the set of commands will be included under a single 'MULTI' operation. Else, specially with commands required to be executed immediately, tracing will happen one-by-one (i.e. the `WATCH` command).
```
Comment (joeblubaugh) [inline README.rst]: Rephrase:
```NOTE: when pipeline commands are executed as a transaction, these commands will be grouped under a single 'MULTI' operation. They'll also appear as a single operation in the trace. Outside of a transaction, each command will generate a span.```

File: examples/explicit.py (hunk 3 above)
Comment (joeblubaugh) [reply examples/explicit.py]: Yes, I would replace this with a comment.

File: redis_opentracing/__init__.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    '''
+    Set our tracer for Redis. Tracer objects from our
```
Comment (joeblubaugh) [inline redis_opentracing/__init__.py]: replace "our django/flask/pyramid" with "the OpenTracing django/flask/pyramid"

File: redis_opentracing/__init__.py (hunk 2 above)
Comment (joeblubaugh) [reply redis_opentracing/__init__.py:112]: Sounds good - this is definitely an evolving part of usage, especially when languages support a wide variety of async models.

**Review by joeblubaugh — APPROVED** [review]
Very nice overall! Some minor changes to wording in a few of the README and comment sections.

**Comment by carlosalberto** [comment]
Merging after applying your suggestions/feedback, Joe. Greatly appreciated. Will poke you on the next iteration (i.e. once we land the `ActiveSpanSource` and integrate it here).
